zeppelin_converter.py

```diff
@@ -0,0 +1,200 @@
+                lang, body = p, None
+
+            lang = lang[1:]
+
+            if lang == 'md':
+                gen_md(lang, body)
+            else:
+                gen_code(lang, body)
+
+            # input_options[lang](lang, body)
+
+        def gen_image(m):
+            result = re.search('base64,(.*)style', m['data'])
+            if result is None:
+                return
+
+            self.index += 1
+
+            if os.path.isdir('images') is False:
+                os.makedirs('images')
+
+            with open("images/output_" + str(self.index) + ".png", "wb") as fh:
+                fh.write(result.group(1).decode('base64'))
+
+            self.print_line("\n![png](images/output_" +
+                            str(self.index) + ".png)\n")
+
+        def gen_text(m):
+            self.print_line(m['data'])
+
+        def create_md_row(r, header=False):
+            cols = r.split('\t')
+            col_md = '|'
+            underline_md = '|'
+
+            if cols != []:
+                for col in cols:
+                    col_md += col + '|'
+                    underline_md += '-|'
+
+            if header:
+                self.print_line(col_md + '\n' + underline_md)
+            else:
+                self.print_line(col_md)
+
+        def gen_table(m):
+            rows = m['data'].split('\n')
+            if rows != []:
+                header_row = rows[0]
+                body_rows = rows[1:]
+                create_md_row(header_row, True)
+                for row in body_rows:
+                    create_md_row(row)
+
+        """
+        Output options routes Zeppelin output types to corresponding
+        functions for it to be handled. To add support for other output
+        types, add the file type to the dictionary and create the necessary
+        function to handle it.
+
+
+        """
+        output_options = {
+            'HTML': gen_image,
+            'TEXT': gen_text,
+            'TABLE': gen_table
+        }
+
+        def process_output(p):
+            if 'editorMode' in p['config']:
+                mode = p['config']['editorMode'].split('/')[-1]
+                if p['results']['msg'] != []:
+                    msg = p['results']['msg'][0]
+                    if mode != 'text' and mode != 'markdown':
+                        output_options[msg['type']](msg)
+
+        def parse_date(d):
+            return datetime.strptime(d, '%b %d, %Y %I:%M:%S %p')
+
+        def process_date_created(t):
+            if self.date_created == 'N/A':
+                self.date_created = t
+            if parse_date(t) < parse_date(self.date_created):
+                self.date_created = t
+
+        def process_date_updated(t):
+            if self.date_updated == 'N/A':
+                self.date_updated = t
+            if parse_date(t) > parse_date(self.date_updated):
+                self.date_updated = t
+
+        for paragraph in t['paragraphs']:
+            if 'user' in paragraph.keys():
```

Comment:
No need to call .keys() here. The target of the `in` operator on a dict are its keys anyway. The same applies to the lines below.